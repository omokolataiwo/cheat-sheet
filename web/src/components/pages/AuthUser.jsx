import React, { Component } from 'react';
import { connect } from 'react-redux';
import Masonry from 'react-masonry-component';
import { fetchFavoriteCheats } from '../../actions/cheat';
import Cheat from '../container/Cheat';

export class AuthUser extends Component {
  state = {};

  componentDidMount() {
    const { fetchFavoriteCheats } = this.props;
    const isAuthorizedUser = localStorage.getItem('token');

    if (isAuthorizedUser) {
      fetchFavoriteCheats();
    }
  }

  componentDidUpdate() {
    const isAuthorizedUser = localStorage.getItem('token');
    const {
      history: { push }
    } = this.props;

    if (!isAuthorizedUser) {
      push('/home');
    }
  }

  renderFavoriteCheat = () => {
    const { cheat: { favorite } } = this.props;

    if (!favorite.length) {
      return (<div className="no-favorite-cheat">You don't have favorite cheat yet.</div>);
    }
    return (
      <Masonry key={new Date()} className="cheat-masonry-grid">
        {localStorage.getItem('token')
        && favorite.map(favorite => (
          <div className="category-section">
            <Cheat cheat={favorite} showActionBar={false} />
          </div>
        ))}
      </Masonry>
    );
  }

  render() {
    const {
      history: { push }
    } = this.props;

    return (
      <div className="container">
        <h4>Favorite Cheats</h4>
        <hr />
        {this.renderFavoriteCheat()}
      </div>
    );
  }
}

const mapStateToProps = ({ user, cheat }) => ({ user, cheat });
export default connect(
  mapStateToProps,
  { fetchFavoriteCheats }
)(AuthUser);
