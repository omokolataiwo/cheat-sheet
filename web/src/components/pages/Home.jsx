import React from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import validate from 'validate.js';
import Masonry from 'react-masonry-component';
import Category from '../container/Category';
import SearchBar from '../container/SearchBar';
import { fetchAllCategoriesWithCheats, addToFavorite, searchSheetCheat } from '../../actions/cheat';
import { signin, signup, resetUserState } from '../../actions/user';
import { SIGNIN_SUCCESSFUL, SIGNOUT_SUCCESSFULLY, SIGNUP_FAILED } from '../../actions/type';
import SideNavPage from '../container/SideNavPage';
import signinValidationConstraint from '../../validatorConstraint/signin';
import signupValidationConstraints from '../../validatorConstraint/signup';

class Home extends React.Component {
  static propTypes = {
    cheat: propTypes.shape({
      cheats: propTypes.arrayOf()
    }).isRequired,
    signin: propTypes.func.isRequired,
    user: propTypes.shape().isRequired,
    history: propTypes.shape({ push: propTypes.func.isRequired }).isRequired
  };

  state = {
    user: {
      firstName: '',
      lastName: '',
      username: '',
      password: ''
    },
    errors: {}
  };

  componentDidUpdate() {
    const {
      user,
      resetUserState,
      history: { push }
    } = this.props;
    if (user.type === SIGNIN_SUCCESSFUL || user.type === SIGNOUT_SUCCESSFULLY) {
      return push('/user');
    }
    if (user.type === SIGNUP_FAILED) {
      resetUserState();
      this.setState(prevState => ({
        errors: { ...prevState.errors, signup: { ...user.errors.signup } }
      }));
    }
  }

  componentWillUnmount() {
    const { searchSheetCheat } = this.props;
    searchSheetCheat('');
  }

  addToFavorite = (id) => {
    const { addToFavorite } = this.props;
    addToFavorite(id);
  };

  formFieldChanged = ({ target: { value, name } }) => {
    this.setState(prevState => ({
      user: { ...prevState.user, [name]: value }
    }));
  };

  signin = (event) => {
    event.preventDefault();
    const { user } = this.state;
    const { signin } = this.props;
    const validationError = validate(user, signinValidationConstraint);

    this.setState(prevState => ({
      errors: { ...prevState.errors, signin: { ...validationError } }
    }));

    if (validationError) {
      return;
    }
    signin(user);
  };

  signup = (event) => {
    event.preventDefault();
    const { user } = this.state;
    const { signup } = this.props;

    const validationError = validate(user, signupValidationConstraints);

    this.setState(prevState => ({
      errors: { ...prevState.errors, signup: { ...validationError } }
    }));

    if (validationError) {
      return;
    }

    signup(user);
  };

  onSearchSheetCheat = (event) => {
    event.preventDefault();
    const { value } = event.target;
    const { searchSheetCheat } = this.props;
    searchSheetCheat(value.trim());
  };

  render() {
    let {
      cheat: { categories, searchResult },
      user: userAction
    } = this.props;
    const { user, errors } = this.state;

    categories = categories || [];
    categories = searchResult || categories;

    return (
      <div className="container">
        <SideNavPage
          onFormFieldChange={this.formFieldChanged}
          user={user}
          errors={errors}
          onSignin={this.signin}
          onSignup={this.signup}
          userAction={userAction}
        />
        <SearchBar onSearchSheetCheat={this.onSearchSheetCheat} />
        <Masonry key={new Date()} className="cheat-masonry-grid">
          {categories.map(category => (
            <Category category={category} onAddToFavorite={this.addToFavorite} />
          ))}
        </Masonry>
      </div>
    );
  }
}

export default connect(
  ({ cheat, user }) => ({ cheat, user }),
  {
    fetchAllCategoriesWithCheats,
    addToFavorite,
    signin,
    signup,
    searchSheetCheat,
    resetUserState
  }
)(Home);
