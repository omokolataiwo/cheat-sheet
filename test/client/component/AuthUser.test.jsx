import React from 'react';
import { shallow } from 'enzyme';
import { AuthUser } from '../../../web/src/components/pages/AuthUser';

const props = {
  history: { push: jest.fn(() => {}) },
  cheat: { type: null, favorite: [] }
};
const wrapper = shallow(<AuthUser {...props} />);

describe('AuthUser Component', () => {
  it('should render component', () => {
    wrapper.setProps({ ...props, fetchFavoriteCheats: jest.fn(() => {}) });
    expect(wrapper).toMatchSnapshot();
  });
  it('should render favorite cheats', () => {
    wrapper.setProps({ ...props, cheat: { type: null, favorite: [{ description: 'Initial repository', line: 'git init' }] } });
  });
});
