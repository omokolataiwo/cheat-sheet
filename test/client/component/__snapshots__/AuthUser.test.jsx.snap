// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthUser Component should render component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthUser
    cheat={
      Object {
        "favorite": Array [],
        "type": null,
      }
    }
    fetchFavoriteCheats={[MockFunction]}
    history={
      Object {
        "push": [MockFunction] {
          "calls": Array [
            Array [
              "/home",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4>
          Favorite Cheats
        </h4>,
        <hr />,
        <div
          className="no-favorite-cheat"
        >
          You don't have favorite cheat yet.
        </div>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Favorite Cheats",
        },
        "ref": null,
        "rendered": "Favorite Cheats",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "You don't have favorite cheat yet.",
          "className": "no-favorite-cheat",
        },
        "ref": null,
        "rendered": "You don't have favorite cheat yet.",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
            Favorite Cheats
          </h4>,
          <hr />,
          <div
            className="no-favorite-cheat"
          >
            You don't have favorite cheat yet.
          </div>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Favorite Cheats",
          },
          "ref": null,
          "rendered": "Favorite Cheats",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "You don't have favorite cheat yet.",
            "className": "no-favorite-cheat",
          },
          "ref": null,
          "rendered": "You don't have favorite cheat yet.",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
